the router implementation simplification

saledjenic opened the pull request:
This PR slightly changes the logic for calculating the best route based on the provided parameters.

To not diverge a lot from the current router logic (to avoid possibly hard-breaking changes) I've tried to improve the following things only:
- New file `router_v2.go` is added that way the logic of the current router is the same as it was.
- Some functions are here just cause we need them for the current router and we will remove them once we're sure that the new logic works fine. An example: we don't need `suggestedFees` function anymore, what it does is now covered with the new `getBaseFee` and `getPriorityFees` functions, but it's still there because of the current router.
- For calculating toke and bonder fees and amount out (an amount that will be received on the destination address) in the case of Hop bridge we don't use our logic (which was a try to replicate Hop's logic on our side which is a very complex one) anymore, but we call a dedicated API endpoint provided by Hop to get all those details.
- The old function `Can` from the `Bridge` interface is now named `AvailableFor` (also `Bridge` name for that interface is not the best one, we should rename it, but it dates from the time when it was used only for bridge txs, later other txs types were added).
- `Path` struct is updated with new property names:
```
type PathV2 struct {
	BridgeName     string
	From           *params.Network    // Source chain
	To             *params.Network    // Destination chain
	FromToken      *walletToken.Token // Token on the source chain
	AmountIn       *hexutil.Big       // Amount that will be sent from the source chain
	AmountInLocked bool               // Is the amount locked
	AmountOut      *hexutil.Big       // Amount that will be received on the destination chain

	SuggestedPriorityFees *PriorityFees // Suggested priority fees for the transaction

	TxBaseFee     *hexutil.Big // Base fee for the transaction
	TxPriorityFee *hexutil.Big // Priority fee for the transaction, by default we're using the Medium priority fee
	TxGasAmount   uint64       // Gas used for the transaction
	TxBonderFees  *hexutil.Big // Bonder fees for the transaction - used for Hop bridge
	TxTokenFees   *hexutil.Big // Token fees for the transaction - used for bridges (represent the difference between the amount in and the amount out)
	TxL1Fee       *hexutil.Big // L1 fee for the transaction - used for for transactions placed on L2 chains

	ApprovalRequired        bool            // Is approval required for the transaction
	ApprovalAmountRequired  *hexutil.Big    // Amount required for the approval transaction
	ApprovalContractAddress *common.Address // Address of the contract that needs to be approved
	ApprovalBaseFee         *hexutil.Big    // Base fee for the approval transaction
	ApprovalPriorityFee     *hexutil.Big    // Priority fee for the approval transaction
	ApprovalGasAmount       uint64          // Gas used for the approval transaction
	ApprovalL1Fee           *hexutil.Big    // L1 fee for the approval transaction - used for for transactions placed on L2 chains

	EstimatedTime TransactionEstimation

	requiredTokenBalance  *big.Int
	requiredNativeBalance *big.Int
}
```
- The router doesn't return any cost (via paths) in any currency, all returned values are in Wei (not GWei).
- The router builds candidates (PathV2) even if there is no balance on the “send from” address and after that it figures out what route (list of paths) is the best one based on the price and that's the route we will return to a client, but also we will check for the possibility of execution of such route and let the client know if the route can be executed based on balances for the selected address on each of the chains from the list.


Basically, those are the initial changes in this PR, I haven't tested them yet, but I wanted firstly to discuss with you is that ok and do we want to continue in that direction?

Since the old `Path` struct is changed and it corresponds to the new `PathV2` struct, that change is a breaking change cause it requires handling on the client side. But I see that now we have a more granular fee structure and it will be up to the client to shape it in the form it wants to display to a user.

That way we can cover complex routes like this:
<img width="481" alt="image" src="https://github.com/status-im/status-go/assets/86303051/994302ca-9b0c-436e-a414-de4c1d4a966e">

<img width="774" alt="image" src="https://github.com/status-im/status-go/assets/86303051/a2e0e032-b6e3-4e28-be27-6279e97020a9">

saledjenic committed:
chore_: `SendType` moved to another file

saledjenic committed:
chore_: the router implementation simplification

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (20)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 3f671f88 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/1/) | 2024-05-14 19:20:13 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 3f671f8813169f1474514eaa00ef8fdede4c1b85 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5159/1/) | 2024-05-14 19:21:43 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240514-191704-3f671f8813169f1474514eaa00ef8fdede4c1b85-pr5159.zip) |
| :heavy_check_mark: | 3f671f8813169f1474514eaa00ef8fdede4c1b85 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5159/1/) | 2024-05-14 19:22:03 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240514-191704-3f671f8813169f1474514eaa00ef8fdede4c1b85-pr5159.zip) |
| :heavy_check_mark: | 3f671f8813169f1474514eaa00ef8fdede4c1b85 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5159/1/) | 2024-05-14 19:23:12 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240514-191704-3f671f8813169f1474514eaa00ef8fdede4c1b85-pr5159.aar) |
| | | | | | | |
| :heavy_check_mark: | daca967af0dfdede8bb869f1c3f1fd594e492a47 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5159/2/) | 2024-05-15 07:32:10 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240515-072927-daca967af0dfdede8bb869f1c3f1fd594e492a47-pr5159.zip) |
| :heavy_check_mark: | daca967af0dfdede8bb869f1c3f1fd594e492a47 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5159/2/) | 2024-05-15 07:32:24 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240515-072927-daca967af0dfdede8bb869f1c3f1fd594e492a47-pr5159.aar) |
| :heavy_check_mark: | daca967af0dfdede8bb869f1c3f1fd594e492a47 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5159/2/) | 2024-05-15 07:33:05 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240515-072927-daca967af0dfdede8bb869f1c3f1fd594e492a47-pr5159.zip) |
| | | | | | | |
| :heavy_check_mark: | daca967a | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/2/) | 2024-05-15 08:12:51 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 7ee491fbe2cc2c4db97faf23961b6ea5d403ca00 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5159/3/) | 2024-05-16 09:02:15 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240516-085903-7ee491fbe2cc2c4db97faf23961b6ea5d403ca00-pr5159.zip) |
| :heavy_check_mark: | 7ee491fbe2cc2c4db97faf23961b6ea5d403ca00 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5159/3/) | 2024-05-16 09:02:34 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240516-085859-7ee491fbe2cc2c4db97faf23961b6ea5d403ca00-pr5159.aar) |
| :heavy_check_mark: | 7ee491fbe2cc2c4db97faf23961b6ea5d403ca00 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5159/3/) | 2024-05-16 09:02:46 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240516-085903-7ee491fbe2cc2c4db97faf23961b6ea5d403ca00-pr5159.zip) |
| | | | | | | |
| :heavy_check_mark: | 7ee491fb | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/3/) | 2024-05-16 09:44:30 | ~45 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 33e1efb3d8848843203322943a9b9d72ae1d9cd8 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5159/4/) | 2024-05-16 10:26:38 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240516-102355-33e1efb3d8848843203322943a9b9d72ae1d9cd8-pr5159.aar) |
| :heavy_check_mark: | 33e1efb3d8848843203322943a9b9d72ae1d9cd8 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5159/4/) | 2024-05-16 10:27:25 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240516-102355-33e1efb3d8848843203322943a9b9d72ae1d9cd8-pr5159.zip) |
| :heavy_check_mark: | 33e1efb3d8848843203322943a9b9d72ae1d9cd8 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5159/4/) | 2024-05-16 10:28:35 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240516-102355-33e1efb3d8848843203322943a9b9d72ae1d9cd8-pr5159.zip) |
| | | | | | | |
| :heavy_check_mark: | 33e1efb3 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/4/) | 2024-05-16 11:08:27 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 5deb408ce091adcb4eb9ba0af80d85d560c4a181 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5159/5/) | 2024-05-17 09:05:46 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240517-090248-5deb408ce091adcb4eb9ba0af80d85d560c4a181-pr5159.zip) |
| :heavy_check_mark: | 5deb408ce091adcb4eb9ba0af80d85d560c4a181 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5159/5/) | 2024-05-17 09:05:52 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240517-090248-5deb408ce091adcb4eb9ba0af80d85d560c4a181-pr5159.aar) |
| :heavy_check_mark: | 5deb408ce091adcb4eb9ba0af80d85d560c4a181 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5159/5/) | 2024-05-17 09:06:27 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240517-090248-5deb408ce091adcb4eb9ba0af80d85d560c4a181-pr5159.zip) |
| | | | | | | |
| :heavy_check_mark: | 5deb408c | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/5/) | 2024-05-17 09:47:03 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/5//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | f7bc3d4d29d33a301034b1e02fed04e4653b10cc | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5159/6/) | 2024-05-17 20:47:32 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240517-204448-f7bc3d4d29d33a301034b1e02fed04e4653b10cc-pr5159.zip) |
| :heavy_check_mark: | f7bc3d4d29d33a301034b1e02fed04e4653b10cc | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5159/6/) | 2024-05-17 20:47:40 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240517-204448-f7bc3d4d29d33a301034b1e02fed04e4653b10cc-pr5159.aar) |
| :heavy_check_mark: | f7bc3d4d29d33a301034b1e02fed04e4653b10cc | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5159/6/) | 2024-05-17 20:48:19 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240517-204448-f7bc3d4d29d33a301034b1e02fed04e4653b10cc-pr5159.zip) |
| | | | | | | |
| :heavy_check_mark: | f7bc3d4d | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/6/) | 2024-05-17 21:28:24 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5159/6//consoleText) |

saledjenic commented:
Things that are not covered here be we should have them in mind are:
- update fees based on base fee and gas price change
- support swap via paraswap where we need to place an approval tx first and wait for its execution and then the swap transaction (that will be handled in `CreateMultiTransactionFromCommand` function)

alaibe commented on services/wallet/bridge/hop.go:
```diff
@@ -109,27 +113,16 @@ func (h *HopBridge) Name() string {
-		return false, nil
-	}
-
+func (h *HopBridge) AvailableFor(from, to *params.Network, token *token.Token, toToken *token.Token) (bool, error) {
```
how do u check that hop supports the token?

saledjenic replied:
That api we're calling `https://api.hop.exchange/v1/quote` will return an like this:
```
{
    "error": "Cannot destructure property 'name' of 'config_1.metadata.tokens[canonicalSymbol]' as it is undefined."
}
```
if the hope doesn't support selected token.

alaibe replied:
alright it means it is handle when computing fees right?

saledjenic replied:
No, it's part of the router logic:
```
can, err := bridge.AvailableFor(network, dest, token, toToken)
if err != nil || !can {
	continue
}
```
means it Hop bridge doesn't support certain token we just skip the Hop bridge as potential candidate for the final route.

alaibe replied:
if i am seeing correclty in bridge.AvailableFor for hop,  you are not looking at token anymore just network right?
if that is the case then the check for supported token is later? in fees computation?

saledjenic replied:
Correct, `CalculateFees` call will check that.

alaibe commented on services/wallet/bridge/hop.go:
```diff
@@ -349,142 +342,50 @@ func (h *HopBridge) swapAndSend(chainID uint64, hopArgs *HopTxArgs, signerFn bin
-	if err != nil {
-		return nil, err
-	}
+func (h *HopBridge) CalculateFees(from, to *params.Network, token *token.Token, amountIn *big.Int) (*big.Int, *big.Int, error) {
```
Those changes will also impact router v1, which is fine, did u ensure that it returns something consistent with [hop](https://hop.exchange/) ?

saledjenic replied:
They won't cause I removed these params `nativeTokenPrice, tokenPrice float64, gasPrice *big.Float` and they were used only for fees calculation for Hop, but now since we're calling hop API for that we don't need them. For other types that implement `Bridge` interface those params were already ignored.

alaibe replied:
I meant the body of the function, not the signature

saledjenic replied:
At the end of that function it returns the token fee and bonder fee, the same as it was, we've just changed the way how we calculate them (now we're fetching them from the remote API). But all that will be undergone thorough checking.

alaibe commented on services/wallet/bridge/hop.go:
```diff
@@ -349,142 +342,50 @@ func (h *HopBridge) swapAndSend(chainID uint64, hopArgs *HopTxArgs, signerFn bin
+	if from.ChainID == walletCommon.OptimismMainnet {
+		fromChainName = HopOptimismChainName
+	} else if from.ChainID == walletCommon.ArbitrumMainnet {
+		fromChainName = HopArbitrumChainName
```
nit: make it a map and avoid if else if else if ;)

saledjenic replied:
I agree, will do.

alaibe commented on services/wallet/fees.go:
```diff
@@ -23,6 +24,9 @@ const (
 )
 
+// //////////////////////////////////////////////////////////////////////////////
+// TODO: remove `SuggestedFees` struct once new router is in place
```
It is not only used by the router, so other places probably needs update

saledjenic replied:
Ahh I see, on the statusgo side it's used only by the router, but I see it's also exposed as an API point. 
Yes, that will require related updates on the client side for the parts that are using it (if those parts do not switch to using the router).

alaibe commented on services/wallet/fees.go:
```diff
@@ -163,6 +176,57 @@ func (f *FeeManager) suggestedFees(ctx context.Context, chainID uint64) (*Sugges
 	}, nil
 }
 
+func (f *FeeManager) getBaseFee(ctx context.Context, client chain.ClientInterface, testnetMode bool) (*big.Int, error) {
```
Shall you update: suggestedFees in the meantime with the new function u created?

saledjenic replied:
I counted that we will remove it sooner or later, but yes, will update it since this new one has a different chain config in the case of testnet.

alaibe commented on services/wallet/router.go:
```diff
@@ -694,22 +529,22 @@ func (r *Router) suggestedRoutes(
 						continue
 					}
 
-					if !sendType.isAvailableBetween(network, dest) {
```
why removing this check?

saledjenic replied:
This check is implemented in `AvailableFor` function of each type individually.

alaibe replied:
After second thought:
```
func (s SendType) isAvailableBetween(from, to *params.Network) bool {
	if s.IsCollectiblesTransfer() {
		return from.ChainID == to.ChainID
	}

	if s == Bridge {
		return from.ChainID != to.ChainID
	}

	if s == Swap {
		return from.ChainID == to.ChainID
	}

	return true
}
```

This was use to prevent path to be created if send type prevent it. e.g u want to do a bridge only then you should only allow from and dest  that are different only. Are u still checking that, I cannot see it?

saledjenic replied:
Yes, all conditions you see here are actually covered and they are the first check in corresponding `AvailableFor` function implementations for each type that implements `Bridge` interface.

For example in case of Hop bridge that implementation looks like this:
```
func (h *HopBridge) AvailableFor(from, to *params.Network, token *token.Token, toToken *token.Token) (bool, error) {
	if from.ChainID == to.ChainID {
		return false, nil
	}
	.
	.
	.
	return true, nil
}
```

In case of Celar bridge looks like this:
```
func (s *CBridge) AvailableFor(from, to *params.Network, token *token.Token, toToken *token.Token) (bool, error) {
	if from.ChainID == to.ChainID {
		return false, nil
	}
	.
	.
	.
	return true, nil
}
```

alaibe commented on services/wallet/router_send_type.go:
```diff
@@ -0,0 +1,169 @@
+package wallet
```
nit: create a router package?

alaibe commented on services/wallet/router_v2.go:
```diff
@@ -0,0 +1,572 @@
+	var best []*PathV2
+	bestCost := big.NewFloat(math.Inf(1))
+	for _, route := range routes {
+		currentCost := big.NewFloat(0)
```
Maybe add a cost function to the PathV2 Type?

saledjenic replied:
I was thinking about it, but my intention was to avoid fetching prices when preparing candidate paths. 
We can easily do that anytime.

alaibe commented on services/wallet/router_v2.go:
```diff
@@ -0,0 +1,572 @@
+package wallet
```
is the logic of this file mostly a copy/paste from v1 adapted?

saledjenic replied:
What I did, I literally copied `router.go` file and then went through the router logic and reorganized some things taking care to not miss anything that we already have there. Guess this way is also easier to compare with the current router.

alaibe commented on services/wallet/router_v2.go:
```diff
@@ -0,0 +1,572 @@
+				if path.ApprovalBaseFee != nil && path.ApprovalPriorityFee != nil {
+					feePerGas := new(big.Int).Add(path.ApprovalBaseFee.ToInt(), path.ApprovalPriorityFee.ToInt())
+					txFeeInWei := new(big.Int).Mul(feePerGas, big.NewInt(int64(path.ApprovalGasAmount)))
+					txFeeInEth := gweiToEth(weiToGwei(txFeeInWei))
```
why are u not returning fees directly in wei?

saledjenic replied:
All fees are in Wei and they are returned to the client in Wei via PathV2, but here we're converting them to `Eth` (Wei to GWei, then GWei to Eth) just to multiply the value with the Eth price to get the final price in Eth that we can compare in order to find the cheapest route.

saledjenic commented:
@alaibe thanks for the review and your comments, they are appreciated since you're the most familiar with the router related code.

There is for sure space for further improvements here, but before we continue I would like to hear from others and be sure that we all have the same vision about it and aim in the same direction.

cammellos commented on services/wallet/api.go:
```diff
@@ -445,6 +445,25 @@ func (api *API) GetSuggestedRoutes(
 }
 
+func (api *API) GetSuggestedRoutesV2(
+	ctx context.Context,
```
if this is a public API, I would pass a request object (i.e a map), so that it's easier to identify which one is the parameter etc, minor thing though

saledjenic replied:
@cammellos just to confirm did you mean to have an API endpoint like this, where keys will be an upfront defined set of params or?
`func (api *API) GetSuggestedRoutesV2(ctx context.Context, params map[string]interface{})(*SuggestedRoutesV2, error)`

cammellos replied:
not quite, I would use types:

```
type GetSuggestedRoutesV2Request struct {
  SendType SendType  `json:"sendType"`
  AddrFrom Commont.Address `json:"addrFrom"`
  ...
} 
```

Much the same as we do with `protocol/requests`
So it's immediately clear what this endpoints arguments are from the request and there's no risk of getting it wrong with positional parameters of the same type

Samyoul commented on services/wallet/bridge/cbridge.go:
```diff
@@ -5,9 +5,9 @@ import (
-	"io/ioutil"
 	"math/big"
-	"net/http"
+	netUrl "net/url"
```
Why alias this import? I can't see a conflict with any other package import.

saledjenic replied:
You're absolutely right, while I was on it I had many other imports there and probably added because of some conflicts, but at the end when I cleaned up the code a bit, forgot to remove it.

Samyoul commented on services/wallet/bridge/cbridge.go:
```diff
@@ -60,43 +63,27 @@ func (s *CBridge) Name() string {
+	params.Add("slippage_tolerance", "500")
 
-	resp, err := client.Do(req)
+	url := fmt.Sprintf("%s/v2/estimateAmt", base)
+	response, err := s.httpClient.DoGetRequest(context.Background(), url, params)
```
Were you avoiding vars shadowing the `url` import with the `netUrl` alias? If yes, the convention (but not the rule) is to change the variable name, rather than the import alias.

Shadowing in Go is allowed, but many devs (myself included) avoid it like you have.

My opinion is that It is fine to make no changes here, just to point out that import aliasing is normally done to handle multiple package imports with the same name, and communicates that the dev should be careful when choosing which package they are selecting.

Samyoul commented on services/wallet/bridge/cbridge.go:
```diff
@@ -112,36 +99,18 @@ func (s *CBridge) getTransferConfig(isTest bool) (*cbridge.GetTransferConfigsRes
 	url := fmt.Sprintf("%s/v2/getTransferConfigs", base)
-
-	req, err := http.NewRequest(http.MethodGet, url, nil)
+	response, err := s.httpClient.DoGetRequest(context.Background(), url, nil)
```
This is a nice change for the context.

Samyoul commented on services/wallet/bridge/hop.go:
Instead of deleting all the logic and depending entirely on hop's API, do you think that we could create a new function that handles the API calls? Just so we don't lose entirely the logic @alaibe implemented here.

We can even use the API calls to validate logic currently present.

saledjenic replied:
Yes, this is more a question for @alaibe, based on what he shared with me about what he was doing there, was a reverse engineering of Hop implementation and if they introduce any change to this logic will be hard for us to know about it. If we ever need it we can dig through git history.

alaibe replied:
My implementation was done before this endpoint existed, I wanted to switch to it so it is fine getting rid of it IMO

Samyoul commented on services/wallet/bridge/swap_paraswap.go:
```diff
@@ -87,7 +87,7 @@ func (s *SwapParaswap) Can(from, to *params.Network, token *walletToken.Token, t
 }
 
-func (s *SwapParaswap) CalculateFees(from, to *params.Network, token *token.Token, amountIn *big.Int, nativeTokenPrice, tokenPrice float64, gasPrice *big.Float) (*big.Int, *big.Int, error) {
+func (s *SwapParaswap) CalculateFees(from, to *params.Network, token *token.Token, amountIn *big.Int) (*big.Int, *big.Int, error) {
```
Has pricing been removed because we use gwei as the common price?

saledjenic replied:
No, those params that are removed are not needed anymore, cause we needed them only to calculate bonder and token fee on our side, now we have them as a response of Hop's api.

Samyoul commented on services/wallet/fees.go:
```diff
@@ -163,6 +176,57 @@ func (f *FeeManager) suggestedFees(ctx context.Context, chainID uint64) (*Sugges
+	}
+
+	// Calculate Low priority fee
+	priorityFee.Low = fees[int64(0.1*float64(len(fees)))-1]
```
How does counting the number of fees help determine the fee value?

`fees, err := f.getFeeHistorySorted(client.NetworkID())` gets historic fees sorted, but `len(fees)` just counts the number of fees, it doesn't handle the fee values.

saledjenic replied:
This logic was taken from `suggestedFees` function. I guess it's about 10th percentile, so we sort fees in an ascending way and take one that 10% of fees are under it, as the the low priority fee.
But if I'm mistaken @alaibe can explain better.

Samyoul replied:
Oh, I see. Yes, thank you.

alaibe replied:
correct it is 10th percentile

Samyoul commented on services/wallet/router_send_type.go:
```diff
@@ -0,0 +1,169 @@
+	return s != ENSRegister && s != ENSRelease && s != ENSSetPubKey && s != StickersBuy
+}
+
+func (s SendType) canUseBridge(b bridge.Bridge) bool {
+	if s == ERC721Transfer && b.Name() != ERC721TransferString {
+		return false
+	}
+
+	if s != ERC721Transfer && b.Name() == ERC721TransferString {
+		return false
+	}
+
+	if s == ERC1155Transfer && b.Name() != ERC1155TransferString {
+		return false
+	}
+
+	if s != ERC1155Transfer && b.Name() == ERC1155TransferString {
+		return false
+	}
+
+	return true
+}
```
Could I make a couple of suggestions for this function:

1. Compatibility map approach:

Map each `SendType` to the expected bridge name. This way, the function looks up the required bridge name for a given `SendType` and compares it directly with the bridge’s name, this reduces the number of conditional statements.

```suggestion
func (s SendType) canUseBridge(b bridge.Bridge) bool {
    compatibilityMap := map[SendType]string{
        ERC721Transfer:  ERC721TransferString,
        ERC1155Transfer: ERC1155TransferString,
    }

    expectedBridge, ok := compatibilityMap[s]
    return !ok || b.Name() == expectedBridge
}
```

2. Switch approach

Directly map `SendType` to the condition without repeating the negative checks. 

```suggestion
func (s SendType) canUseBridge(b bridge.Bridge) bool {
    bridgeName := b.Name()
    switch s {
    case ERC721Transfer:
        return bridgeName == ERC721TransferString
    case ERC1155Transfer:
        return bridgeName == ERC1155TransferString
    default:
        return true // Assumes other types can use any bridge
    }
}
```

Both approaches have their advantages.

Samyoul commented on services/wallet/router_send_type.go:
```diff
@@ -0,0 +1,169 @@
+	return true
+}
+
+func (s SendType) isAvailableFor(network *params.Network) bool {
+	if s == Swap {
+		return network.ChainID == walletCommon.EthereumMainnet ||
+			network.ChainID == walletCommon.OptimismMainnet ||
+			network.ChainID == walletCommon.ArbitrumMainnet
+	}
+
+	if s == Transfer || s == Bridge || s.IsCollectiblesTransfer() {
+		return true
+	}
+
+	if network.ChainID == walletCommon.EthereumMainnet ||
+		network.ChainID == walletCommon.EthereumGoerli ||
+		network.ChainID == walletCommon.EthereumSepolia {
+		return true
+	}
+
+	return false
+}
```
This function has a number of compound conditionals, so perhaps we could declare the conditions first and then compare after. Something like:

```go
func (s SendType) isAvailableFor(network *params.Network) bool {
    // Set of network ChainIDs allowed for any type of transaction
    allAllowedNetworks := map[uint64]bool{
        walletCommon.EthereumMainnet: true,
        walletCommon.EthereumGoerli:  true,
        walletCommon.EthereumSepolia: true,
    }

    // Additional specific networks for the Swap SendType
    swapAllowedNetworks := map[uint64]bool{
        walletCommon.EthereumMainnet:  true,
        walletCommon.OptimismMainnet:  true,
        walletCommon.ArbitrumMainnet:  true,
    }

    // Check for Swap specific networks
    if s == Swap {
        return swapAllowedNetworks[network.ChainID]
    }

    // Check for any SendType available for all networks
    if s == Transfer || s == Bridge || s.IsCollectiblesTransfer() || allAllowedNetworks[network.ChainID] {
        return true
    }

    return false
}
```

This approach makes it more clear what the networks are used for what purpose.

Samyoul commented on services/wallet/router_send_type.go:
```diff
@@ -0,0 +1,169 @@
+	return false
+}
+
+func (s SendType) EstimateGas(service *Service, network *params.Network, from common.Address, tokenID string) uint64 {
+	tx := transactions.SendTxArgs{
+		From:  (types.Address)(from),
+		Value: (*hexutil.Big)(zero),
+	}
+	if s == ENSRegister {
+		estimate, err := service.ens.API().RegisterEstimate(context.Background(), network.ChainID, tx, EstimateUsername, EstimatePubKey)
+		if err != nil {
+			return 400000
+		}
+		return estimate
+	}
+
+	if s == ENSRelease {
+		estimate, err := service.ens.API().ReleaseEstimate(context.Background(), network.ChainID, tx, EstimateUsername)
+		if err != nil {
+			return 200000
+		}
+		return estimate
+	}
+
+	if s == ENSSetPubKey {
+		estimate, err := service.ens.API().SetPubKeyEstimate(context.Background(), network.ChainID, tx, fmt.Sprint(EstimateUsername, ".stateofus.eth"), EstimatePubKey)
+		if err != nil {
+			return 400000
+		}
+		return estimate
+	}
```
Perhaps this is a good case for a `switch` statement.

saledjenic committed:
chore_: review comments applied (part1)

saledjenic committed:
chore_: router moved to router package (part2)

J-Son89 commented on services/wallet/api.go:
```diff
@@ -445,6 +445,25 @@ func (api *API) GetSuggestedRoutes(
+	ctx context.Context,
+	sendType SendType,
+	addrFrom common.Address,
+	addrTo common.Address,
```
I see a lot of these parameters already exists so I don't know easy a change this is.
However, I find parameter/variable names that drop letters to be confusing. 
Sure we save a few characters but it's generally just better to use full words so the developer doesn't have to guess what the meaning is (not everyone is a native speaker) and on the otherside you almost have to guess what letters have been dropped as this is often an inconsistent and random choice. 
Overall it also leaves the codebase cleaner imo and that serves a lot in the long run. ⚔️

saledjenic commented:
I addressed all your comments. If you're ok with that, we can merge these changes and I can continue inspecting each sending type individually and applying fixes if needed?

J-Son89 commented on services/wallet/router/fees.go:
```diff
@@ -23,6 +25,9 @@ const (
 )
 
+// //////////////////////////////////////////////////////////////////////////////
+// TODO: remove `SuggestedFees` struct once new router is in place
```
can we include a GITHUB issue for this TODO - In Mobile that's how we have been doing it as it makes it easier to find out if the relevant issue was dropped or updated in any sense afterwards.

J-Son89 commented on services/wallet/router/fees.go:
```diff
@@ -89,7 +100,13 @@ func gweiToWei(val *big.Float) *big.Int {
 
-func (f *FeeManager) suggestedFees(ctx context.Context, chainID uint64) (*SuggestedFees, error) {
+// //////////////////////////////////////////////////////////////////////////////
+// TODO: remove `suggestedFees` function once new router is in place
```
same here

J-Son89 commented on services/wallet/bridge/cbridge.go:
```diff
@@ -60,43 +63,27 @@ func (s *CBridge) Name() string {
-		return nil, err
-	}
+	params := url.Values{}
+	params.Add("src_chain_id", strconv.Itoa(int(from.ChainID)))
```
same for these, can't we just add full name params? :)
Why are some using this approach and some not? consistency adds a lot of value imo

saledjenic replied:
`src_chain_id ` this is not something that we defined, this is the param requested by clear API and it must be like that.

J-Son89 replied:
Yep but the many cases below this too

saledjenic committed:
chore_: checks per sending type added (part3)

alaibe reviewed (approved):
LGTM nice update

briansztamfater commented:
> Since the old `Path` struct is changed and it corresponds to the new `PathV2` struct, that change is a breaking change cause it requires handling on the client side. But I see that now we have a more granular fee structure and it will be up to the client to shape it in the form it wants to display to a user.

Is there a simple way to see the diff between the old and new struct so it is easier to identify the changes to be made in the client side?

saledjenic commented:
> > Since the old `Path` struct is changed and it corresponds to the new `PathV2` struct, that change is a breaking change cause it requires handling on the client side. But I see that now we have a more granular fee structure and it will be up to the client to shape it in the form it wants to display to a user.
> 
> Is there a simple way to see the diff between the old and new struct so it is easier to identify the changes to be made in the client side?

It will be explained how to migrate once we confirm that it works. 
The only changes for reponse will be in details about the fees.

Samyoul commented on services/wallet/router/router.go:
```diff
@@ -437,22 +277,41 @@ func newSuggestedRoutes(
+	erc1155Transfer := bridge.NewERC1155TransferBridge(rpcClient, transactor)
+	cbridge := bridge.NewCbridge(rpcClient, transactor, tokenManager)
+	hop := bridge.NewHopBridge(rpcClient, transactor, tokenManager)
+	paraswap := bridge.NewSwapParaswap(rpcClient, transactor, tokenManager)
```
I'm sorry that I missed this previously. I notice that you've simplified the bridge inits by considerably increasing the `NewRouter()` init function's parameter complexity. I think that it is fine to pass in a single struct with multiple complex fields.

The Options pattern can help in this case or a config pattern. But also the original approach was fine. What problem does this change solve? Perhaps I'm not noticing something.

saledjenic replied:
@Samyoul That was an easy way to solve the cyclic dependency, that I ran into while moving the router to a new `router` package.

Without this the `router.go` (which is the `router` package) is importing the package `wallet` and `wallet/api.go` is importing `router` package while it's part of `wallet` package, that results in cyclic dependency.

Samyoul replied:
Oh I see. Ok. I won't spend too much brain power on this then. Thank you.

Samyoul reviewed (approved):
Thank you for the changes, and for all the work you've done to get this far. I've approved this PR with one minor point about function signature complexity (which is a point of opinion rather than a critical issue).
